### PR TITLE
feat(gates): structured per-angle findings in the verdict comment (#898)

### LINKED ISSUE #898: Gate verdict comment should render consolidated per-angle fan-in findings structurally, not collapsed free text
## Summary

The gate verdict comment (`renderGateReviewCommentBody` in `scripts/github/upsert-checkpoint-verdict.mjs`) templates the *frame* well — `Gate review` / `Reviewed head SHA` / `Verdict` / `Execution mode` / `Findings summary:` / `Next action`. But the **findings-summary body is operator free text**, and when passed via `--findings-summary` it is run through `summarizeCheckpointVerdictText()` → `collapseWhitespace()`, which flattens it to a single run-on line (newlines only survive via `--findings-file`).

Result: a `fanout_fanin` verdict — which consolidates N per-angle reviews — renders as a wall of text (`correctness: … contract-surface: … acceptance-criteria: …`) with no structure. Shape depends entirely on operator discipline, so it routinely comes out unreadable (every fanout_fanin verdict this session).

## The real fix

For `fanout_fanin`, the body should NOT be hand-written free text. The fan-in already produces **structured per-angle data** (`consolidateFanin` / `toFindingsLogShape` in `packages/core/src/loop/gate-fanin.mjs`: per-angle verdict + findings + severity). The verdict renderer should **render that structure** deterministically — e.g. one line/bullet per angle with its verdict and any findings — driven by the consolidated object, not a free-text arg.

Concretely:
- Add a structured findings input to `upsert-checkpoint-verdict` (e.g. accept the consolidated fan-in JSON, or read the findings-log ledger for the head) and render per-angle bullets in `renderGateReviewCommentBody`.
- Keep free-text `--findings-summary`/`--findings-file` as a fallback for `inline_single_agent`.
- Ensure newlines/markdown bullets survive (don't `collapseWhitespace` the structured render).

## Acceptance

- [x] A `fanout_fanin` verdict comment renders consolidated per-angle findings as a readable structured list (verdict + findings per angle), template-driven from the fan-in data.
- [x] Newlines/bullets are preserved (not collapsed).
- [x] `inline_single_agent` free-text path still works.
- [x] Test asserting the structured render shape.

## Provenance

Surfaced 2026-06-26 by the owner reviewing the #897 draft_gate verdict comment (run-on single line). Part of the gate-contract rework alongside #895/#896.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by mfittko**
## Summary

The gate verdict comment (`renderGateReviewCommentBody` in `scripts/github/upsert-checkpoint-verdict.mjs`) rendered the findings-summary as operator free text, and `--findings-summary` collapses it to a single run-on line. For a `fanout_fanin` verdict that consolidates N per-angle reviews, this produced an unreadable wall of text.

This PR renders the consolidated per-angle fan-in findings **structurally**, driven by structured data instead of operator discipline.

## Input mechanism

A new optional `--findings-json <path>` flag carries the structured fan-out review findings. The **primary** shape is the per-angle review-results array — the same per-angle objects the fan-out reviewers write and that feed `consolidateFanin`'s *input* (`packages/core/src/loop/gate-fanin.mjs`):

```json
[{ "angle": "correctness", "verdict": "findings_present",
   "findings": [{ "severity": "must-fix", "summary": "...", "file": "a.mjs", "line": 7, "disposition": "accepted-for-fix" }] }]
```

It **also** accepts the FLAT per-finding array that `consolidateFanin` / `toFindingsLogShape` *produce* (`[{ "severity", "summary", "angle?", "file?"/"files?", ... }]`); a flat array is **grouped by each finding's `.angle`** into per-angle sections (findings without an `.angle` are grouped under `general`), so a real flat output is rendered correctly instead of being silently dropped. A non-empty input matching **neither** shape (and a mix of the two) is **rejected with a descriptive error** rather than silently rendering an all-clean verdict. It also defensively accepts an object wrapping the array under `angles`/`findings`. When supplied, it takes precedence over `--findings-summary`/`--findings-file` for the rendered body.

## Render shape

One bullet per angle (angle label in a code span + per-angle verdict), with nested findings carrying severity, summary, optional `file:line`, and disposition:

```
**Findings summary:** 2 angles reviewed; 1 finding (see per-angle breakdown below).

- `correctness` → findings_present
  - [must-fix] broken edge case (`a.mjs:7`) — _accepted-for-fix_
- `tests` → clean
```

Newlines are **preserved** (the structured block is NOT run through `collapseWhitespace` / `summarizeCheckpointVerdictText`). The block is bounded by `MAX_GATE_COMMENT_TEXT_LENGTH`. Angle/finding text is sanitized for markdown (per-line whitespace collapse, backtick-strip inside code spans, HTML-comment delimiter neutralization — mirroring `post-gate-findings.mjs`).

## How the marker/parse contract is preserved

- The `**Findings summary:**` line still carries a non-empty **single-line digest** — exactly what `parseGateReviewCommentFields` captures for that field (it only reads the remainder of that one line).
- The structured bullets are deliberately written so no nested line matches a gate field regex (no `verdict:` / `next action:` / `execution mode:` line starts; angle verdict uses `→` not `verdict:`).
- `gate`, `headSha`, `verdict`, `executionMode`, `nextAction` still round-trip and `contractComplete` stays `true`, so `summarizeGateReviewComments` / `detect-checkpoint-evidence` / pre-merge gate checks are unaffected. A test asserts this round-trip on the structured body.

The free-text `--findings-summary` / `--findings-file` path is unchanged as the fallback for `inline_single_agent`.

## Wired end-to-end

The `fanout_fanin` gate-verdict step in `skills/copilot-pr-followup/SKILL.md` (mandatory gate-comment command contract + Phase 4 of the fan-out procedure) now passes `--findings-json <path>` with the per-angle review results, instead of `--findings-summary`. `docs/gate-review-sub-loop-contract.md` documents the same. `--findings-summary` remains documented as the `inline_single_agent` fallback. `.claude/` assets regenerated (`--check` passes).

## Tests

- Structured render produces a readable multi-line per-angle block (verdict + findings), newlines preserved.
- The structured comment round-trips through the marker parser (gate/headSha/verdict/executionMode/nextAction/contractComplete).
- Sanitization: backticks stripped, embedded newlines collapsed, HTML-comment delimiters neutralized.
- End-to-end `--findings-json` posts the structured body and reports `executionMode: fanout_fanin`.
- Nested per-angle input renders correctly; FLAT per-finding input (with `.angle` per finding) renders grouped without dropping findings; an unrecognizable non-empty shape (and a mixed shape) throws / exits 1.
- Free-text fallback for inline verdicts unchanged.
- `npm run verify` exit 0; package-lock churn reverted.

## Checklist

- [x] Structured per-angle render, template-driven from fan-in data
- [x] Newlines/bullets preserved (not collapsed)
- [x] `inline_single_agent` free-text path unchanged
- [x] Marker/parse contract preserved (round-trip test)
- [x] Accepts both per-angle and flat shapes; rejects unrecognizable input (no silent all-clean)
- [x] Wired into the `fanout_fanin` gate-verdict step (SKILL.md + sub-loop contract); `--findings-summary` is the inline fallback
- [x] `npm run verify` green

Closes #898

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by mfittko** [comment]
### Gate review: `draft_gate`

**Reviewed head SHA:** `1ab013b7210bf2b4cc5b487a2dbebc261ec1df36`
**Verdict:** clean
**Execution mode:** fanout_fanin

**Findings summary:** Draft gate clean via fan-out/fan-in (correctness+parse-contract, acceptance-criteria).

- correctness/parse-contract (clean): structured per-angle render round-trips through the gate-evidence parser (bullets can't mis-parse as gate fields; the digest line preserves the Findings-summary field); injection sanitized; free-text fallback intact; length bounded; malformed JSON handled.
- acceptance-criteria (clean after fix): the two integration findings are fixed — normalizeStructuredFindings now accepts per-angle OR flat-grouped shapes and THROWS on unrecognizable input (no silent all-clean), and the fanout_fanin gate-verdict command in the skill is wired to --findings-json so #898's structured render is realized end-to-end. Closes #898.

**Next action:** Mark ready; request Copilot.

File: scripts/github/upsert-checkpoint-verdict.mjs [hunk 1]
```diff
@@ -400,6 +435,231 @@ function appendGateEvidenceNote(summary, note) {
...
+// Sort findings by severity (must-fix first) for deterministic output,
+// preserving input order within a severity.
+function sortStructuredFindings(findings) {
+  findings.sort(
+    (a, b) =>
+      STRUCTURED_FINDINGS_SEVERITY_ORDER.indexOf(a.severity)
+      - STRUCTURED_FINDINGS_SEVERITY_ORDER.indexOf(b.severity),
+  );
+  return findings;
+}
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/upsert-checkpoint-verdict.mjs]: Fixed in bab73967. `sortStructuredFindings()` no longer feeds `indexOf()` directly into the comparator. A new `severitySortRank()` maps known severities to their `STRUCTURED_FINDINGS_SEVERITY_ORDER` index (must-fix → worth-fixing-now → defer) and maps any unknown/missing severity to a large rank (`STRUCTURED_FINDINGS_SEVERITY_ORDER.length`), so unknowns sort LAST instead of floating above must-fix on the old `-1` rank. The sort stays deterministic (stable sort preserves input order within a severity). New test `renderGateReviewCommentBody sorts unknown/missing severities LAST, never before must-fix` asserts must-fix leads and unknown/missing trail.

File: scripts/github/upsert-checkpoint-verdict.mjs [hunk 2]
```diff
@@ -400,6 +435,231 @@ function appendGateEvidenceNote(summary, note) {
...
+  }
+  const recognizable = input.filter((item) => looksLikePerAngleEntry(item) || looksLikeFlatFinding(item));
+  if (recognizable.length === 0) {
+    throw new Error(
+      "--findings-json input is non-empty but matches neither recognized shape: "
+      + "a per-angle array ([{ angle, verdict?, findings: [...] }]) or a flat "
+      + "per-finding array ([{ severity, summary, angle?, ... }]). Refusing to "
+      + "render an all-clean verdict from unrecognized findings.",
+    );
+  }
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/upsert-checkpoint-verdict.mjs]: Fixed in bab73967. `normalizeStructuredFindings()` no longer filters the input down to "recognizable" entries and proceeds. It now computes the unrecognized items and THROWS a descriptive error when a non-empty payload contains ANY item that is neither a recognizable per-angle entry nor a recognizable flat finding — so unrecognized items from producer drift or malformed entries can no longer be silently dropped from a gate verdict. (The all-unrecognized case still throws the existing "matches neither recognized shape" error; the mixed-shape throw is preserved.) New test `renderGateReviewCommentBody throws when a non-empty payload mixes recognized and unrecognized items (no silent drop)` covers a payload with one valid per-angle entry plus one unrecognized item.

File: scripts/github/upsert-checkpoint-verdict.mjs [hunk 3]
```diff
@@ -400,6 +435,231 @@ function appendGateEvidenceNote(summary, note) {
...
+    return false;
+  }
+  return typeof item.summary === "string" && item.summary.trim().length > 0;
+}
+// Build a render-ready per-angle section from a nested entry.
+function buildAngleSectionFromNested(raw) {
+  const angle = typeof raw.angle === "string" ? raw.angle.trim() : "";
+  if (angle.length === 0) {
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/upsert-checkpoint-verdict.mjs]: Fixed in bab73967. `buildAngleSectionFromNested()` no longer returns null on a missing/blank `angle`. It now renders such an entry under a `general` fallback label — consistent with the flat-grouping angleless→`general` bucket — so its findings still appear rather than being dropped. Because the entry is no longer dropped, a non-empty `structuredFindings` payload can no longer silently degrade to the free-text path due to a dropped angle-less section. New test `renderGateReviewCommentBody renders an angle-less NESTED entry under general instead of dropping it` asserts both an angle-missing and a blank-angle nested entry render under `general` and that the structured digest (not the free-text summary) is used.

File: scripts/github/upsert-checkpoint-verdict.mjs [hunk 4]
```diff
@@ -790,13 +1121,19 @@ export async function upsertCheckpointVerdict(options, { env = process.env, ghCo
       throw new Error(`Cannot read --findings-file "${options.findingsFile}": ${err instanceof Error ? err.message : String(err)}`);
     }
   }
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/upsert-checkpoint-verdict.mjs]: Fixed — when `--findings-json` is provided (and takes precedence), `--findings-file` is no longer read, so passing both with a missing/invalid file no longer causes a spurious hard failure. (guard: `if (!structuredFindings && options.findingsFile)`)

File: scripts/github/upsert-checkpoint-verdict.mjs [hunk 5]
```diff
@@ -42,6 +42,29 @@ Required:
                                             alternative to --findings-summary
                                             (preserves newlines; takes precedence
                                             when both are present)
+  --findings-json <path>                    Read STRUCTURED fan-out review
+                                            findings from a JSON file. PRIMARY
+                                            shape is the per-angle review-results
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/upsert-checkpoint-verdict.mjs:48]: Fixed in d42b4dff. The top USAGE synopsis now advertises all three findings inputs — `(--findings-summary <text> | --findings-file <path> | --findings-json <path>)` — and adds a line noting that the `--findings-json` structured per-angle path is preferred for `--execution-mode fanout_fanin`, so `--help` no longer hides the new fanout_fanin path. The per-flag `--findings-json` help block was already present and is unchanged.

File: scripts/github/upsert-checkpoint-verdict.mjs [hunk 6]
```diff
@@ -422,9 +704,26 @@ export function renderGateReviewCommentBody({ gate, headSha, verdict, findingsSu
...
+  // with newlines preserved (NOT collapsed to a run-on line).
+  const angles = normalizeStructuredFindings(structuredFindings);
+  if (angles) {
+    lines.push(
+      "",
+      `**Findings summary:** ${buildStructuredFindingsDigest(angles)}`,
+      "",
+      renderStructuredFindings(angles),
+    );
+  } else {
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/upsert-checkpoint-verdict.mjs:728]: Fixed in d42b4dff. The structured branch of `renderGateReviewCommentBody` no longer ignores the gate-evidence note. It now takes a `gateEvidenceNote` parameter and renders the single-line summary as `appendGateEvidenceNote(buildStructuredFindingsDigest(angles), gateEvidenceNote ?? null)` — the same append helper (and same `; ` separator + `MAX_GATE_COMMENT_TEXT_LENGTH` bound) the free-text path uses. `upsertCheckpointVerdict` passes `coordination.gateEvidenceNote` into the renderer, so a structured verdict's `**Findings summary:**` line now carries notes like the round-cap / round-exhaustion fallback note instead of dropping them. The per-angle bullets are unchanged and the summary stays a single line, so the marker/`parseGateReviewCommentFields` round-trip still passes. New test asserts a structured `--findings-json` verdict whose coordination emits a gateEvidenceNote renders the note on the summary line and still round-trips through the gate parser.

File: scripts/github/upsert-checkpoint-verdict.mjs [hunk 7]
```diff
@@ -790,13 +1124,19 @@ export async function upsertCheckpointVerdict(options, { env = process.env, ghCo
...
-    ? options.findingsSummary
-    : appendGateEvidenceNote(options.findingsSummary, coordination.gateEvidenceNote ?? null);
+  // The findings-summary the comment is compared/round-tripped against. With a
+  // structured render this is the single-line digest (what the marker parser
+  // recovers from the `**Findings summary:**` line); otherwise the free-text path.
+  const effectiveFindingsSummary = structuredFindings
+    ? buildStructuredFindingsDigest(structuredFindings)
+    : (options.findingsFile
+      ? options.findingsSummary
+      : appendGateEvidenceNote(options.findingsSummary, coordination.gateEvidenceNote ?? null));
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/upsert-checkpoint-verdict.mjs:1142]: Fixed in d42b4dff. In structured mode `effectiveFindingsSummary` is now `appendGateEvidenceNote(buildStructuredFindingsDigest(structuredFindings), coordination.gateEvidenceNote ?? null)` instead of the bare digest — parity with the free-text path, so structured verdict comments no longer omit the gate-evidence note. The renderer is given the same `coordination.gateEvidenceNote` and appends it identically, so the rendered `**Findings summary:**` line matches the `effectiveFindingsSummary` used for the noop / round-trip compare, and `appendGateEvidenceNote` keeps it a single length-bounded line so the gate parser still round-trips. Covered by the new structured-with-note parity test.

**Comment by mfittko** [comment]
### Gate review: `pre_approval_gate`

**Reviewed head SHA:** `d42b4dffab698f131e8a2ecadf3d303563fe554d`
**Verdict:** clean
**Execution mode:** fanout_fanin

**Findings summary:** Pre-approval gate clean via fan-out/fan-in (correctness+parse-contract, acceptance-criteria).

- correctness/parse-contract (clean): structured per-angle render round-trips through the gate-evidence parser; injection sanitized; free-text fallback intact. Hardened across review: never silently drops/hides findings (unknown severities sort last; unrecognized payloads throw; angle-less entries render under 'general'); --findings-json precedence skips --findings-file; structured summary carries the gateEvidenceNote (parity with free-text).
- acceptance-criteria (clean): #898 ACs met end-to-end — structured per-angle render, newlines preserved, inline fallback intact, tests; the fanout_fanin gate-verdict command is wired to --findings-json; --findings-json accepts per-angle or flat-grouped shapes. Closes #898.

Copilot review rounds exhausted (4/2); current head has zero unresolved threads and green or credibly green CI, so pre_approval_gate fallback is allowed without another Copilot re-request.

**Next action:** Auto-merge: squash-merge.
